### PR TITLE
monodepth_midas

### PR DESCRIPTION
**Pull request by dvdhfnr**
tech-transfer: mono-depth - midas
- model
- demo
- accuracy checker

INFO: downloader is still not able to download our model weights from our gdrive:
python downloader.py --name midasnet
################|| Downloading models ||################

========== Downloading /code/open_model_zoo/tools/downloader/public/midasnet/midasnet.py
... 100%, 6 KB, 16721 KB/s, 0 seconds passed

========== Downloading /code/open_model_zoo/tools/downloader/public/midasnet/midasnet.pth
... 0%, 3 KB, 9308 KB/s, 0 seconds passed
########## Error: Downloaded file is shorter (3243 B) than expected (422388453 B)

################|| Post-processing ||################

FAILED:
midasnet

### CODE REVIEW

File: tools/accuracy_checker/configs/midasnet.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
+models:
+  - name: midasnet
+    launchers:
+      - framework: dlsdk
+        tags:
+          - FP32
+        # path to IR relative to open_model_zoo/models for FP32/FP16 converted precisions
```
Comment (eaidova) [inline tools/accuracy_checker/configs/midasnet.yml]: please remove this comment

File: tools/accuracy_checker/configs/midasnet.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        model: /code/open_model_zoo/tools/downloader/public/midasnet/FP32/midasnet.xml
```
Comment (eaidova) [inline tools/accuracy_checker/configs/midasnet.yml]: We use path relative to open_model_zoo/models directory
```suggestion
        model: public/midasnet/FP32/midasnet.xml
```

Comment (dvdhfnr) [reply tools/accuracy_checker/configs/midasnet.yml]: done

File: tools/accuracy_checker/configs/midasnet.yml (hunk 1 above)
Comment (dvdhfnr) [reply tools/accuracy_checker/configs/midasnet.yml]: done

File: tools/downloader/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+public/
```
Comment (eaidova) [inline tools/downloader/.gitignore]: Is it pushed intentionally?

File: tools/accuracy_checker/configs/midasnet.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+models:
+  - name: midasnet
+    launchers:
+      - framework: dlsdk
+        tags:
+          - FP32
+        model: public/midasnet/FP32/midasnet.xml
+        weights: public/midasnet/FP32/midasnet.bin
+        adapter: mono_depth
```
Comment (eaidova) [inline tools/accuracy_checker/configs/midasnet.yml]: could you please also provide launcher for FP16 model here?
We need it for validation. It is very simillar to FP32:
```yml
      - framework: dlsdk
        tags:
          - FP16
        model: public/midasnet/FP16/midasnet.xml
        weights: public/midasnet/FP16/midasnet.bin
        adapter: mono_depth

File: tools/downloader/.gitignore (hunk 3 above)
Comment (dvdhfnr) [reply tools/downloader/.gitignore]: removed

File: tools/accuracy_checker/configs/midasnet.yml (hunk 4 above)
Comment (dvdhfnr) [reply tools/accuracy_checker/configs/midasnet.yml]: added

File: demos/python_demos/monodepth_demo/.gitignore [hunk 5]
```diff
@@ -0,0 +1,3 @@
+disp.png
```
Comment (eaidova) [inline demos/python_demos/monodepth_demo/.gitignore]: this also should be removed, I suppose

Comment (dvdhfnr) [reply demos/python_demos/monodepth_demo/.gitignore]: done

File: models/public/midasnet/model.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    sha256: a3680912f1e6839b5bc4d4684e78b2a8f5f669125ee86aff81b53dbe8d635a12
+    source:
+      $type: google_drive
+      id: 1fL_UDsZ8pFUN3QnXqMF4BReVZmfmrKBj
+  - name: midasnet.pth
+    size: 422388453
+    sha256: f46da7438f3dd46a46fa550cea1d5dcab4bb3192ba272f1fd2b08d23ba41d39e
+    source:
+      $type: google_drive
+      id: 1m6-zyyiZj2XCiemSdQ7ldrEW-E1ZFfIl
```
Comment (IRDonch) [inline models/public/midasnet/model.yml]: This appears to be the same file that is linked from the repository, but it's been reuploaded under a different ID. Why reupload it?

File: models/public/midasnet/model.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+  The model output is an inverse depth map that is defined up to an unknown scale factor.
+task_type: mono_depth
+files:
+  - name: midasnet.py
+    size: 6464
+    sha256: a3680912f1e6839b5bc4d4684e78b2a8f5f669125ee86aff81b53dbe8d635a12
+    source:
+      $type: google_drive
+      id: 1fL_UDsZ8pFUN3QnXqMF4BReVZmfmrKBj
```
Comment (IRDonch) [inline models/public/midasnet/model.yml]: Is this code different in any way from what's in the repository?

**Comment by dvdhfnr** [comment]
Referring to the comment of eizamaliev:
Actually, we do *not* need to download any external model weights. This is not used by the inference code. It was a relict from the past. I changed our midasnet.py file on our gdrive accordingly. We never need to download those weights (use_pretrained = False). Now, there is no downloading in the conversion of the model to onnx / dldt.

File: models/public/midasnet/model.yml (hunk 7 above)
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: The code is different since our base model can not be converted to dldt via onnx. Some changes were required to the original code; such as fixed image size, changes in upsampling.

File: models/public/midasnet/model.yml (hunk 6 above)
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: See my comment below: We needed to apply changes to our original code to make the model convertible to dltl (via onnx). We want to keep the weights and (adapted) model definition consistent. That is why we keep them in the same gdrive folder.

File: tools/downloader/requirements-pytorch.in [hunk 8]
```diff
@@ -1,4 +1,4 @@
 onnx
 scipy           # via torchvision
-torch
+torch==1.1      # required to convert upsample_bilinear2d
```
Comment (eizamaliev) [inline tools/downloader/requirements-pytorch.in]: The model can be converted with all PyTorch, except 1.2. Could you test it with >1.2?

**Comment by eizamaliev** [comment]
> Referring to the comment of eizamaliev:
> *skip*

The described problems are solved with latest fix. 
Now I want ask you to test conversion, see https://github.com/opencv/open_model_zoo/pull/843#discussion_r378302626

File: models/public/midasnet/midasnet.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
+# midasnet
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]: Please, add information about I/O, and summary about model. See [here](https://github.com/opencv/open_model_zoo/blob/develop/models/public/alexnet/alexnet.md) for example.

**Comment by eizamaliev** [comment]
Also, see our [Contribution Guide](https://github.com/opencv/open_model_zoo/blob/develop/CONTRIBUTING.md)

File: tools/downloader/requirements-pytorch.in (hunk 8 above)
Comment (dvdhfnr) [reply tools/downloader/requirements-pytorch.in]: requires torch>=1.4
I update the req. file accordingly.

File: demos/python_demos/monodepth_demo/README.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# MonoDepth Python Demo
```
[the remark this answers is not in the document]
Comment (dvdhfnr) [reply demos/python_demos/monodepth_demo/README.md]: - updated demos' root README.md
- added models.lst

￼

File: models/public/midasnet/midasnet.md (hunk 9 above)
Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: readme updated.

**Comment by dvdhfnr** [comment]
> Also, see our [Contribution Guide](https://github.com/opencv/open_model_zoo/blob/develop/CONTRIBUTING.md)

I tried to follow those guidelines.

File: tools/downloader/requirements.in [hunk 11]
```diff
@@ -1,2 +1,3 @@
 pyyaml
 requests
+test-generator
```
Comment (eizamaliev) [inline tools/downloader/requirements.in]: Why did you add this package here? In fact, I couldn't find its usage in your code.

File: models/public/midasnet/midasnet.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+The model input is a blob that consists of a single image of "1x3x384x384" in `BGR` order.
+
+The model output is an inverse depth map that is defined up to an unknown scale factor.
+
+## Example
+
+See [here](https://github.com/intel-isl/MiDaS)
+
+## Input
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]: Usually, we describe inputs and outputs for both original and converted in IR models, even if they are same. Once again, see [as example](https://github.com/opencv/open_model_zoo/blob/develop/models/public/alexnet/alexnet.md).

File: models/public/midasnet/midasnet.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+- `B` - batch size
+- `H` - height
+- `W` - width
+
+Inverse depth map is defined up to an unknown scale factor.
+
+## Legal Information
+
+[LICENSE](https://drive.google.com/open?id=1p_7P7VKSpD1xM8Ex6p0epZ4TdYFPYjss)
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]: Since last release, we provide not only link to license, but full license text here too.

File: models/public/midasnet/midasnet.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+<https://arxiv.org/pdf/1907.01341>
+
+The model input is a blob that consists of a single image of "1x3x384x384" in `BGR` order.
+
+The model output is an inverse depth map that is defined up to an unknown scale factor.
+
+## Example
+
+See [here](https://github.com/intel-isl/MiDaS)
+
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]: Please, add `Specification` section here (see [example](https://github.com/opencv/open_model_zoo/blob/develop/models/public/alexnet/alexnet.md#specification)).

**Comment by eizamaliev** [comment]
Also, I would ask you to add your model to [index](https://github.com/opencv/open_model_zoo/blob/develop/models/public/index.md). And since model sovles new task, it would be nice to see some description of the task.

File: tools/downloader/requirements.in (hunk 11 above)
Comment (dvdhfnr) [reply tools/downloader/requirements.in]: It was a missing req when running the converter. Hoeever, you are right, it should be part of the dldt req. Removed.

File: models/public/midasnet/midasnet.md (hunk 14 above)
Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: Ok. Is there an easy way to determine the GFLOPs and Mparams in OpenVino?

File: models/public/midasnet/midasnet.md (hunk 13 above)
Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: done

File: models/public/midasnet/midasnet.md (hunk 12 above)
Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: done

**Comment by dvdhfnr** [comment]
> Also, I would ask you to add your model to [index](https://github.com/opencv/open_model_zoo/blob/develop/models/public/index.md). And since model sovles new task, it would be nice to see some description of the task.

done.

File: models/public/midasnet/midasnet.md (hunk 14 above)
Comment (eizamaliev) [reply models/public/midasnet/midasnet.md]: You can use Model Analyzer tool via DL Workbench, or manually (see `<openvino_install_dir>/deployment_tools/tools/workbench/model_analyzer` directory)

Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: Is this also part of the dldt repo? I did not install the openvino_toolkit so far.

File: models/public/midasnet/midasnet.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+### Original Model
+
+Image, name - `data`, shape - `1,3,384,384`, format is `B,C,H,W` where:
+
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `RGB`.
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]: Please, add here mean and scale value if any. But only for *original model*.

The main idea of this section is to provide information about which input is expected, and what kind of preprocessing should be done.

Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: man and scale values added

File: models/public/midasnet/midasnet.md (hunk 14 above)
Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: Might be a silly question: I build dltl from the repo. I cannot find the model_analyzer. Do I have do enable a CMAKE flag or is it not includes in the repo?

Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: It was required to do this since I was earlier constrained to the master branch of dldt. This dependency is now gone with the 2020 version. Hence, I can go ahead with the toolkit...

Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: I am able to report the GFLOPs. However, it returns MParams: 0
Looking into the code shows: 
    def get_total_params(self) -> int:
        raise NotImplementedError

I would go ahead with just commenting on the GFLOPs

File: models/public/midasnet/model.yml [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+description: >-
+  MidasNet is a model for monocular depth estimation trained by mixing several datasets;
+  as described in the following paper: 
+  "Towards Robust Monocular Depth Estimation: Mixing Datasets for Zero-Shot Cross-Dataset Transfer"
+  <https://arxiv.org/pdf/1907.01341>
+
+  The model input is a blob that consists of a single image of "1x3x384x384" in RGB order.
+
+  The model output is an inverse depth map that is defined up to an unknown scale factor.
+task_type: mono_depth
```
[the remark this answers is not in the document]
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: Depth estimation would be too general. There are several depth estimation approaches, e.g. stereo, time-of-flight, etc. And single-image based. In the literature, latter one is referred to as mono-depth (or monocular depth estimation). This is a well-established name.

File: demos/python_demos/monodepth_demo/monodepth_demo.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    disp = cv2.resize(disp, (input_width, input_height), cv2.INTER_CUBIC)
+
+    # rescale disp
+    disp_min = disp.min()
+    disp_max = disp.max()
+
+    if disp_max - disp_min > 1e-6:
+        disp = (disp - disp_min) / (disp_max - disp_min)
+    else:
+        disp = 0.5
```
[the remark this answers is not in the document]
Comment (dvdhfnr) [reply demos/python_demos/monodepth_demo/monodepth_demo.py]: You are correct. Thanks. Corrected.

File: demos/README.md [hunk 18]
```diff
@@ -30,6 +30,7 @@ The Open Model Zoo includes the following demos:
 - [Super Resolution C++ Demo](./super_resolution_demo/README.md) - Super Resolution demo (the demo supports only images as inputs). It enhances the resolution of the input image.
 - [Text Detection C++ Demo](./text_detection_demo/README.md) - Text Detection demo. It detects and recognizes multi-oriented scene text on an input image and puts a bounding box around detected area.
 - [Text Spotting Python* Demo](./python_demos/text_spotting_demo/README.md) - The demo demonstrates how to run Text Spotting models.
+- [Monodepth Python* Demo](./python_demos/monodepth_demo/README.md) - The demo demonstrates how to run monocular depth estimation models.
```
[the remark this answers is not in the document]
Comment (dvdhfnr) [reply demos/README.md]: ok.

File: tools/downloader/common.py [hunk 19]
```diff
@@ -57,6 +57,7 @@
     'object_attributes',
     'optical_character_recognition',
     'semantic_segmentation',
+    'mono_depth',
```
[the remark this answers is not in the document]
Comment (dvdhfnr) [reply tools/downloader/common.py]: done.

File: models/public/midasnet/model.yml (hunk 7 above)
Comment (IRDonch) [reply models/public/midasnet/model.yml]: I took a look at the changes you made to the original code. Other than concatenation, there were four changes:

1. Replacing `torch.hub.load` with a torchvision model.
2. Replacing scale factors for interpolation with sizes (and specifying explicit sizes everywhere).
3. Replacing `align_corners=True` with `align_corners=False` in `FeatureFusionBlock`.
4. Replacing `use_pretrained = False if path else True` with just `use_pretrained = False`.

Now, I've managed to convert the model by taking the original code and applying changes 1, 3 and 4. Change 2 doesn't actually seem necessary. Maybe it was only needed for older PyTorches and/or Model Optimizers?

Changes 3 and 4 are simple enough that you can do them through Model Downloader itself (just search for `regex_replace` in other model configs to see how this can be done). Change 1 looks like you could probably just apply it in the original repository, but if you don't want to, then you can also do it with `regex_replace`.

All this means that you can change the model config to get the sources from the original repository, and I would prefer if you did so. This would be better for a few reasons:

1. It would make it clear where the code came from to someone looking at the config.
2. It would explicitly enumerate the changes that have to be made to the code in order to convert it to the IR format.
3. Change 4 might become unnecessary in the future if Model Optimizer supports ONNX opset 11, in which case it would be prudent to remove it. This will be a lot easier to do if it's applied via a post-processing step during downloading.

File: models/public/midasnet/model.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (IRDonch) [reply models/public/midasnet/model.yml]: OK, how about `monocular_depth_estimation`, then? That is at least consistent with the other task types.

File: models/public/midasnet/midasnet.md (hunk 14 above)
Comment (eizamaliev) [reply models/public/midasnet/midasnet.md]: Fine, but remove `?` sign from MParams row

File: models/public/midasnet/midasnet.md [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+Image, name - `image`, shape - `1,3,384,384`, format is `B,C,H,W` where:
+
+- `B` - batch size
+- `C` - channel
+- `H` - height
+- `W` - width
+
+Channel order is `RGB`.
+
+- `mean_values` - `image[123.675, 116.28, 103.53]`
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]:
```suggestion
Mean values - [123.675, 116.28, 103.53].
```

File: models/public/midasnet/midasnet.md [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+- `scale_values` - `image[51.525, 50.4, 50.625]`
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]:
```suggestion
Scale values - [51.525, 50.4, 50.625].
```

File: models/public/midasnet/midasnet.md (hunk 14 above)
Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: done.

File: models/public/midasnet/model.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: ok, sure. done.

File: models/public/midasnet/midasnet.md [hunk 22]
```diff
@@ -0,0 +1,104 @@
+# midasnet
+
+## Use Case and High-Level Description
+
+MidasNet is a model for monocular depth estimation trained by mixing several datasets;
+as described in the following paper:
+"Towards Robust Monocular Depth Estimation: Mixing Datasets for Zero-Shot Cross-Dataset Transfer"
+<https://arxiv.org/pdf/1907.01341>
+
+The model input is a blob that consists of a single image of "1x3x384x384" in `BGR` order.
```
Comment (eizamaliev) [inline models/public/midasnet/midasnet.md]:
```suggestion
The model input is a blob that consists of a single image of "1x3x384x384" in `RGB` order.
```
Because it is description of original model.

Comment (dvdhfnr) [reply models/public/midasnet/midasnet.md]: Original or converted model?
original: RGB
converted: BGR

File: models/public/midasnet/model.yml (hunk 7 above)
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: I would klike to debug this, but the download of the weight from the gdrive is still not working; see https://github.com/opencv/open_model_zoo/pull/843#issue-373737647

Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: I tested the downloading / conversion with a local webserver. Works fine. I commited the changes. Remainign issue: downloading from gdrive fails.

Comment (IRDonch) [reply models/public/midasnet/model.yml]: I actually can't download it myself either anymore without logging into Google. Google just says that the file has been downloaded too many times and is now unavailable. Not sure we can do anything about this but wait...

(That said, for debugging the downloading of the model _code_ you could probably just comment out the weights file in the config and place the weights file into the model directory manually.)

Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: Downloading via chrome works fine for me. 
However, downloading via the downloader tool not.

Maybe at makes sense to test the download tool w/ and w/o being logged-in to google.

Comment (IRDonch) [reply models/public/midasnet/model.yml]: > Downloading via chrome works fine for me.

Even in an incognito window?

> It might make sense to test the download tool w/ and w/o being logged-in to google.

The tool can't itself log in, and your browser login shouldn't make a difference.

**Comment by IRDonch** [comment]
Jenkins: test this please

File: models/public/midasnet/model.yml [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+
+  The model input is a blob that consists of a single image of "1x3x384x384" in RGB order.
+
+  The model output is an inverse depth map that is defined up to an unknown scale factor.
+task_type: monocular_depth_estimation
+files:
+  - name: midas_net.py
+    size: 2662
+    sha256: 0bf2e5295b69b3af9adafaf97e154186d29a808325435440aa68c7aa4d25f753
+    source: https://raw.githubusercontent.com/intel-isl/MiDaS/master/models/midas_net.py
```
Comment (IRDonch) [inline models/public/midasnet/model.yml]: Please use a specific commit hash rather than `master`. (Otherwise, there's no guarantee that the version we want will be fetched.)

File: models/public/midasnet/model.yml (hunk 7 above)
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: That is correct. Then, I do not understand why the download via the tool is/was not working for you, but for me.

File: models/public/midasnet/model.yml (hunk 23 above)
Comment (dvdhfnr) [reply models/public/midasnet/model.yml]: done.

File: models/public/midasnet/model.yml (hunk 7 above)
Comment (IRDonch) [reply models/public/midasnet/model.yml]: It's a good question, but now it works for neither of us, so it's kinda of moot. :-) For now I'm just going to wait 24 hours as the Google Drive message suggests and see if the situation improves.

(I'm also going to resolve this conversation, since my original comment is no longer relevant.)

**Comment by IRDonch** [comment]
There are a few Pylint watnings (http://134.191.240.124/job/omz/job/precommit/job/accuracy_checker_ubuntu16/2120/console). Please fix them.

**Comment by dvdhfnr** [comment]
> There are a few Pylint warnings (http://134.191.240.124/job/omz/job/precommit/job/accuracy_checker_ubuntu16/2120/console). Please fix them.

done.

**Comment by dvdhfnr** [comment]
@IRDonch The model.pt is now again downloadable via chrome (also in incognito mode).
Download via the omz downloader not working (at least for me).

**Comment by IRDonch** [comment]
Jenkins: test this please
